File: cheapestFlightsWIthKStopsDijkstra.java
import java.util.*;

class Solution {
    public int findCheapestPrice(int n, int[][] flights, int src, int dst, int k) {
        // adjacency list for storing the graph
        List<List<int[]>> adj = new ArrayList<>();
        for (int i = 0; i < n; i++) {
            adj.add(new ArrayList<>());
        }

        // prepare the adjacency list
        for (int[] e : flights) {
            adj.get(e[0]).add(new int[]{e[1], e[2]});
        }

        // stops array to count the number
        int[] stops = new int[n];
        Arrays.fill(stops, Integer.MAX_VALUE);

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> a[0] - b[0]);
        pq.offer(new int[]{0, src, 0});

        while (!pq.isEmpty()) {
            int[] temp = pq.poll();
            int dist = temp[0];
            int node = temp[1];
            int steps = temp[2];

            if (steps > stops[node] || steps > k + 1) continue;
            stops[node] = steps;

            if (node == dst) return dist;

            for (int[] neighbor : adj.get(node)) {
                pq.offer(new int[]{dist + neighbor[1], neighbor[0], steps + 1});
            }
        }

        return -1;
    }
}

 

            


           

File: cheapestFlightsWithKstopsDijkstra.py
import heapq
from typing import List
from math import inf

class Solution:
    def findCheapestPrice(self, n: int, flights: List[List[int]], src: int, dst: int, k: int) -> int:
        # adjacency list for storing the graph
        adj = [[] for _ in range(n)]

        # prepare the adjacency list
        for e in flights:
            adj[e[0]].append((e[1], e[2]))

        # stops array to count the number
        stops = [inf] * n

        pq = [(0, src, 0)]

        while pq:
            dist, node, steps = heapq.heappop(pq)

            if steps > stops[node] or steps > k + 1:
                continue

            stops[node] = steps

            if node == dst:
                return dist

            for neighbor, price in adj[node]:
                heapq.heappush(pq, (dist + price, neighbor, steps + 1))

        return -1
